app/Http/Controllers/MonitorController.php

```diff
@@ -0,0 +1,121 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Http\Controllers;
+
+require_once 'include/common.php';
+
+use DateTime;
+use DateInterval;
+use DatePeriod;
+use App\Enums\ClassicPalette;
+use App\Enums\HighContrastPalette;
+use Illuminate\Http\JsonResponse;
+use Illuminate\Support\Facades\Auth;
+use Illuminate\Support\Facades\DB;
+use Symfony\Component\HttpFoundation\Response;
+use Illuminate\Http\Request;
+
+class MonitorController extends AbstractController
+{
+    /**
+     * Get statistics about our recent submissions queue performance.
+     */
+    public function get(): JsonResponse
+    {
+        $user = Auth::user();
+        if ($user === null || !$user->IsAdmin()) {
+            return response()->json(['error' => 'Permissions error'], status: Response::HTTP_FORBIDDEN);
+        }
+
+        // Get the length of the current backlog and how long the oldest submission
+        // has been waiting.
+        $backlog_length = \DB::table('jobs')->count();
+        $backlog_time = null;
+        if ($backlog_length > 0) {
+            $backlog_time = time_difference(time() - \DB::table('jobs')->min('created_at'), false, 'ago');
+        }
+
+        // Choose colors for the chart based on user preferences.
+        if (array_key_exists('colorblind', $_COOKIE) && intval($_COOKIE['colorblind']) === 1) {
+            $palette = HighContrastPalette::class;
+        } else {
+            $palette = ClassicPalette::class;
+        }
+
+        // Initialize trendline data.
+        $success_values = [];
+        $fail_values = [];
+        $ticks = [];
+        $period = new \DatePeriod(
+            new DateTime('24 hours ago'),
+            new DateInterval('PT1H'),
+            24);
+
+        $i = 0;
+        foreach ($period as $datetime) {
+            $timestamp = $this->truncateTimestampToHour($datetime);
+            $success_values[$timestamp] = 0;
+            $fail_values[$timestamp] = 0;
+            $i++;
+            if ($i % 3 === 0) {
+                $ticks[] = $timestamp;
+            }
+        }
+
+        // Populate pass/fail trendline data.
+        foreach (\DB::table('failed_jobs')->pluck('failed_at') as $failed_at) {
+            $key = $this->truncateTimestampToHour(new DateTime($failed_at));
```

Comment:
This is going to potentially result in a *lot* of rows being returned.  Between this query and the query on line 76, tens or hundreds of thousands of rows could be returned on large production systems.  A better solution might be to perform the timestamp truncation on the DB side via MySQL's `hour()` or PostgreSQL's `date_part()`, grouping by the resulting hour and using the `count` aggregate to get the number of results for that hour.  Doing this will result in a maximum of 24 rows being returned, which is much more manageable.  The downside of this approach is that there is no single SQL function which can do this.  MySQL and PostgreSQL each have their own functions, which means that you would have to perform a bit of additional logic to determine which database is being used.  I'll leave the decision regarding whether this tradeoff is worthwhile to your judgement.